Redirect after login/registration to the desired URL

ash211 opened the pull request.

ash211 committed:
Redirect after login/registration to the desired URL

thelastnode commented:
GET parameters instead? Django does it that way.

Also, we don't have stale data: if they log out and log back in, it doesn't redirect them to the last page that had a loginRequired decorator.

ash211 commented:
I went the session path because it was originally suggested by bnoguchi at https://github.com/bnoguchi/everyauth/issues/41

You're right though, GET parameters have a lot of advantages.  They also handle the case with multiple tabs open on a single page, which then allows you to get back to all your desired pages after, say, logging into LAWN.

ash211 committed:
Updated method for login redirect on success

1) user clicks on restricted page /private
2) redirected to /login?next=0%2Fprivate
3) the next variable is hidden in the login form
4) on submit, the next URL sent to server
5) if valid, redirect to the received URL

ash211 commented:
How's the new method look Ankit?

thelastnode commented on app.js:
```diff
@@ -16,7 +16,28 @@ everyauth.password
+        };
+    })
+    .respondToRegistrationSucceed( function (res, user, data) {
+        if (data.req && data.req.body && data.req.body.next) {
```
When would `data.req` not exist? For a request, `req.body` always exists, even if it's empty. We can get rid of a couple conditions here if that's the case.

thelastnode commented on app.js:
```diff
@@ -16,7 +16,28 @@ everyauth.password
+            next: req.query.next
+        };
+    })
+    .respondToLoginSucceed( function (res, user, data) {
```
Whitespace convention:

```
.respondToLoginSucceed(function(req, user, data) {
```

thelastnode commented on app.js:
```diff
@@ -16,7 +16,28 @@ everyauth.password
+            res.end();
+        }
+    })
+    .registerLocals( function (req, res) {
```
Let's move these functions into `auth.js`, with the rest of the authentication functions.

thelastnode commented on app.js:
```diff
@@ -16,7 +16,28 @@ everyauth.password
+    })
+    .respondToLoginSucceed( function (res, user, data) {
+        if (user && data.req && data.req.body && data.req.body.next) {
+            res.writeHead(303, {'Location': data.req.body.next});
```
We always need to redirect, not only if `data.req.body.next` exists, so redirect to `/` by default.

thelastnode commented:
@ash211 ping?

ash211 committed:
Move auth handling functions into lib/auth.js

ash211 committed:
Add default redirect location: '/'

ash211 committed:
Remove unused registerSuccessRedirect call

ash211 committed:
Eliminate unneeded variable existence checks

ash211 committed:
Fix whitespace

thelastnode commented on lib/auth.js:
```diff
@@ -103,5 +104,28 @@ exports.loginRequired = function(req, res, next) {
+};
+
+exports.respondToLoginSucceed = function(res, user, data) {
+    if(!user) return;
```
If authentication has failed, we should redirect to the login URL with the next parameter intact, and flash an error.

ash211 replied:
Already being tracked in a separate request, doesn't block merging this one. Note that the wording of this method implies that it only gets called when login actually succeeds, but actually it gets called for both success and failure, surprise! Correct handling should likely be done in respondToLoginFail

https://github.com/cc-wit/attendance/issues/44

thelastnode replied:
Oh, this does handle errors by redirecting to the login page and flashing the errors. Do you know where the flashing of errors is handled? We should at least add a comment saying why this line exists, and where to look to see how errors are handled.

ash211 replied:
I believe it's in the default implementation of respondToLoginFail

thelastnode replied:
Oh, [you're right](https://github.com/bnoguchi/everyauth/blob/master/lib/modules/password.js#L98). We should add a comment mentioning this.

thelastnode commented on lib/auth.js:
```diff
@@ -103,5 +104,28 @@ exports.loginRequired = function(req, res, next) {
+};
+
+exports.respondToRegistrationSucceed = function(res, user, data) {
+    if(!user) return;
```
Same with registration, but we should maintain the inputs that were given (except password). This may be grounds for using [express-form](https://github.com/dandean/express-form), if it works anything like Django's forms.

ash211 replied:
Looks like express-form would do nicely for the frontend validation portion, tracking part of that (non-blank passwords at least) here: https://github.com/cc-wit/attendance/issues/46

thelastnode commented on lib/auth.js:
```diff
@@ -103,5 +104,28 @@ exports.loginRequired = function(req, res, next) {
+
+    var loc = '/';
+    if(data.req.body.next) {
+        loc = data.req.body.next;
```
```
var loc = data.req.body.next || '/'
```

ash211 replied:
Done.

thelastnode commented on views/login.jade:
```diff
@@ -14,6 +14,7 @@ form(action='/login', method='POST')
     label(for='password') Password
     input(type='password', name=everyauth.password.passwordFormFieldName)#password
   p
+    input(type='hidden', name='next', value=next)
```
GET parameter? Moving this to a GET parameter also means using `req.params.data` instead of `req.body.data`, I think.

ash211 replied:
No advantage to going GET immediately, since we're posting already and getting a redirect to the final location.  If it makes it easier to handle persisting destination through failed login attempts, we can make this change in the future.

ash211 committed:
More succinct default handling
